### ISSUE TITLE
Bug in Regex-Engine (?)

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#12949]
Migrated from [rt.perl.org#117917](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=117917) (status was 'resolved')

Searchable as RT117917$

**Comment by p5pRT** [comment]
### From kochnorman@rocketmail.com

I am not sure\, whether perlbug correctly sent this or not\. Sorry\, if 
this is duplicate content to you\.

The attached file is the bugreport\, that I \(hope\) I just sent\. If so\, 
ignore this mail please\.

**Comment by p5pRT** [comment]
### From kochnorman@rocketmail.com

#### Created by kochnorman@rocketmail.com



I was just playing around trying to build a regex for the rather complex problem of matching a q\-like\-operator\-string in Perl \(like qq/hallo/\, but also "hallo" and so on\)\. So\, I came to trying this&#8203;:

use strict;
use warnings;

use re "debugcolor"; 

sub getDel \{
&nbsp;      my $z = shift;
&nbsp;      die $z;
&nbsp;      return '\)' if $z eq '\(';
&nbsp;      return '>' if $z eq '\<';
&nbsp;      return '\]' if $z eq '\[';
&nbsp;      return '\}' if $z eq '\{';
&nbsp;      return $z;
\} 

'qq"hello"' =~ /\(?\<stringstart>\(?&#8203;:q\(?&#8203;:\[xwrq\]\)?\\s\*\)?\)\(?\<delim>\(??\{ \(exists \($\+\{stringstart\}\) && $\+\{stringstart\} =~ m\#q\#\) ? q\#"|'\# : '\\\\w' \}\)\)\.\*?\(?\<\!\\\\\)\(??\{ die $\+\{delim\}; defined $1 ? quotemeta\(getDel\($1\)\) : '\(?\!\)' \}\)/


When executing this code\, I get this output \(including re "debugcolor"\-Output\)&#8203;:

Compiling REx "\(?\<stringstart>\(?&#8203;:q\(?&#8203;:\[xwrq\]\)?\\s\*\)?\)\(?\<delim>\(??\{ \(exists \($"\.\.\.
Compiling REx "q"
Final program&#8203;:
&nbsp;   1&#8203;: EXACT \<q> \(3\)
&nbsp;   3&#8203;: END \(0\)
anchored "q" at 0 \(checking anchored isall\) minlen 1 
Final program&#8203;:
&nbsp;   1&#8203;: OPEN1 'stringstart' \(3\)
&nbsp;   3&#8203;:   CURLYX\[0\] \{0\,1\} \(23\)
&nbsp;   5&#8203;:     EXACT \<q> \(7\)
&nbsp;   7&#8203;:     CURLY \{0\,1\} \(20\)
&nbsp;   9&#8203;:       ANYOF\[qrwx\]\[\] \(0\)
&nbsp;  20&#8203;:     STAR \(22\)
&nbsp;  21&#8203;:       SPACE \(0\)
&nbsp;  22&#8203;:   WHILEM \(0\)
&nbsp;  23&#8203;:   NOTHING \(24\)
&nbsp;  24&#8203;: CLOSE1 'stringstart' \(26\)
&nbsp;  26&#8203;: OPEN2 'delim' \(28\)
&nbsp;  28&#8203;:   LOGICAL\[2\] \(29\)
&nbsp;  29&#8203;:   EVAL \(31\)
&nbsp;  31&#8203;: CLOSE2 'delim' \(33\)
&nbsp;  33&#8203;: MINMOD \(34\)
&nbsp;  34&#8203;: STAR \(36\)
&nbsp;  35&#8203;:   REG\_ANY \(0\)
&nbsp;  36&#8203;: UNLESSM\[\-1\] \(42\)
&nbsp;  38&#8203;:   EXACT \<\\\\> \(40\)
&nbsp;  40&#8203;:   SUCCEED \(0\)
&nbsp;  41&#8203;: TAIL \(42\)
&nbsp;  42&#8203;: LOGICAL\[2\] \(43\)
&nbsp;  43&#8203;: EVAL \(45\)
&nbsp;  45&#8203;: END \(0\)
minlen 0 with eval 
Matching REx "\(?\<stringstart>\(?&#8203;:q\(?&#8203;:\[xwrq\]\)?\\s\*\)?\)\(?\<delim>\(??\{ \(exists \($"\.\.\. against "qq%"hallo%""
&nbsp;   0 \<qq"hallo">|  1&#8203;:OPEN1 'stringstart'\(3\)
&nbsp;   0 \<qq"hallo">|  3&#8203;:CURLYX\[0\] \{0\,1\}\(23\)
&nbsp;   0 \<qq"hallo">| 22&#8203;:  WHILEM\(0\)
&nbsp;                                    whilem&#8203;: matched 0 out of 0\.\.1
&nbsp;   0 \<qq"hallo">|  5&#8203;:    EXACT \<q>\(7\)
&nbsp;   1 \<qq"hallo">|  7&#8203;:    CURLY \{0\,1\}\(20\)
&nbsp;                                      ANYOF\[qrwx\]\[\] can match 1 times out of 1\.\.\.
&nbsp;   2 \<qq"hallo">| 20&#8203;:      STAR\(22\)
&nbsp;                                        SPACE can match 0 times out of 2147483647\.\.\.
&nbsp;   2 \<qq"hallo">| 22&#8203;:        WHILEM\(0\)
&nbsp;                                          whilem&#8203;: matched 1 out of 0\.\.1
&nbsp;   2 \<qq"hallo">| 23&#8203;:          NOTHING\(24\)
&nbsp;   2 \<qq"hallo">| 24&#8203;:          CLOSE1 'stringstart'\(26\)
&nbsp;   2 \<qq"hallo">| 26&#8203;:          OPEN2 'delim'\(28\)
&nbsp;   2 \<qq"hallo">| 28&#8203;:          LOGICAL\[2\]\(29\)
&nbsp;   2 \<qq"hallo">| 29&#8203;:          EVAL\(31\)
Guessing start of match in sv for REx "q" against "qq"
Found anchored substr "q" at offset 0\.\.\.
Guessed&#8203;: match at offset 0
Matching embedded REx "%"|'" against "%"hallo%""
&nbsp;   2 \<qq"hallo">|  1&#8203;:            TRIE\-EXACT\["'\]\(7\)
&nbsp;   2 \<qq"hallo">|                State&#8203;:    1 Accepted&#8203;: N Charid&#8203;:  1 CP&#8203;:  22 After State&#8203;:    2
&nbsp;   3 \<qq"hallo">|                State&#8203;:    2 Accepted&#8203;: Y Charid&#8203;:  0 CP&#8203;:   0 After State&#8203;:    0
&nbsp;                                              got 1 possible matches
&nbsp;                                              TRIE matched word \#1\, continuing
\[1\]    8691 segmentation fault  perl qq\.pl


\.\.\. and I cannot really see why perl crashes here\. I didn't except that regex to work perfectly or even to work at all \(regarding to matching the correct things\)\, because this task is maybe too complex to achieve it with a regex\, but I certainly did not expect perl to crash with a segmentation fault\. 


When disabling re "debugcolor"\, the output is "Use of uninitialized value $\+\{"delim"\} in die at \(re\_eval 2\) line 1\.
Died at \(re\_eval 2\) line 1\."\. Probably re cannot handle that by now?

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=medium

Site configuration information for perl 5.14.2:

Configured by Debian Project at Fri Apr 12 09:56:36 UTC 2013.

Summary of my perl5 (revision 5 version 14 subversion 2) configuration:
   
  Platform:
    osname=linux, osvers=2.6.32-5-686-bigmem, archname=i486-linux-gnu-thread-multi-64int
    uname='linux murphy 2.6.32-5-686-bigmem #1 smp mon feb 25 01:53:47 utc 2013 i686 gnulinux '
    config_args='-Dusethreads -Duselargefiles -Dccflags=-DDEBIAN -D_FORTIFY_SOURCE=2 -g -O2 -fstack-protector --param=ssp-buffer-size=4 -Wformat -Werror=format-security -Dldflags= -Wl,-z,relro -Dlddlflags=-shared -Wl,-z,relro -Dcccdlflags=-fPIC -Darchname=i486-linux-gnu -Dprefix=/usr -Dprivlib=/usr/share/perl/5.14 -Darchlib=/usr/lib/perl/5.14 -Dvendorprefix=/usr -Dvendorlib=/usr/share/perl5 -Dvendorarch=/usr/lib/perl5 -Dsiteprefix=/usr/local -Dsitelib=/usr/local/share/perl/5.14.2 -Dsitearch=/usr/local/lib/perl/5.14.2 -Dman1dir=/usr/share/man/man1 -Dman3dir=/usr/share/man/man3 -Dsiteman1dir=/usr/local/man/man1 -Dsiteman3dir=/usr/local/man/man3 -Duse64bitint -Dman1ext=1 -Dman3ext=3perl -Dpager=/usr/bin/sensible-pager -Uafs -Ud_csh -Ud_ualarm -Uusesfio -Uusenm -Ui_libutil -DDEBUGGING=-g -Doptimize=-O2 -Duseshrplib -Dlibperl=libperl.so.5.14.2 -des'
    hint=recommended, useposix=true, d_sigaction=define
    useithreads=define, usemultiplicity=define
    useperlio=define, d_sfio=undef, uselargefiles=define, usesocks=undef
    use64bitint=define, use64bitall=undef, uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cc', ccflags ='-D_REENTRANT -D_GNU_SOURCE -DDEBIAN -fstack-protector -fno-strict-aliasing -pipe -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-O2 -g',
    cppflags='-D_REENTRANT -D_GNU_SOURCE -DDEBIAN -fstack-protector -fno-strict-aliasing -pipe -I/usr/local/include'
    ccversion='', gccversion='4.7.2', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=12345678
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long long', ivsize=8, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=4, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -fstack-protector -L/usr/local/lib'
    libpth=/usr/local/lib /lib/i386-linux-gnu /lib/../lib /usr/lib/i386-linux-gnu /usr/lib/../lib /lib /usr/lib
    libs=-lgdbm -lgdbm_compat -ldb -ldl -lm -lpthread -lc -lcrypt
    perllibs=-ldl -lm -lpthread -lc -lcrypt
    libc=, so=so, useshrplib=true, libperl=libperl.so.5.14.2
    gnulibc_version='2.13'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-Wl,-E'
    cccdlflags='-fPIC', lddlflags='-shared -L/usr/local/lib -fstack-protector'

Locally applied patches:
    


@INC for perl 5.14.2:
    /usr/share/perl/5.14.2
    /usr/share/perl
    /etc/perl
    /usr/local/lib/perl/5.14.2
    /usr/local/share/perl/5.14.2
    /usr/lib/perl5
    /usr/share/perl5
    /usr/lib/perl/5.14
    /usr/share/perl/5.14
    /usr/local/lib/site_perl
    .


Environment for perl 5.14.2:
    HOME=/home/nok
    LANG=de_DE.UTF-8
    LANGUAGE=
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/home/nok/.vim/bin:/usr/local/bin:/usr/bin:/bin:/usr/bin/X11:/usr/games
    PERL5LIB=/usr/share/perl
    PERL_BADLANG (unset)
    PERL_LOCAL_LIB_ROOT=/home/nok/perl5
    PERL_MB_OPT=--install_base /home/nok/perl5
    PERL_MM_OPT=INSTALL_BASE=/home/nok/perl5
    SHELL=/usr/bin/zsh

```
</details>

**Comment by p5pRT** [comment]
### From @jkeenan

On Wed May 08 13&#8203;:36&#8203;:33 2013\, kochnorman@&#8203;rocketmail\.com wrote&#8203;:
> I am not sure\, whether perlbug correctly sent this or not\. Sorry\, if 
> this is duplicate content to you\.
> 
> The attached file is the bugreport\, that I \(hope\) I just sent\. If so\, 
> ignore this mail please\.
> 
> 

The program as I have extracted it from your attachment&#8203;:

\#\#\#\#\#
$ cat 117917\_regex\.pl 
use strict;
use warnings;

use re "debugcolor"; 

sub getDel \{
&nbsp;          my $z = shift;
&nbsp;          die $z;
&nbsp;          return '\)' if $z eq '\(';
&nbsp;          return '>' if $z eq '\<';
&nbsp;          return '\]' if $z eq '\[';
&nbsp;          return '\}' if $z eq '\{';
&nbsp;          return $z;
\} 

'qq"hello"' =~ /\(?\<stringstart>\(?&#8203;:q\(?&#8203;:\[xwrq\]\)?\\s\*\)?\)\(?\<delim>\(??\{
\(exists \($\+\{stringstart\}\) && $\+\{stringstart\} =~ m\#q\#\) ? q\#"|'\# : '\\\\w'
\}\)\)\.\*?\(?\<\!\\\\\)\(??\{ die $\+\{delim\}; defined $1 ? quotemeta\(getDel\($1\)\) :
'\(?\!\)' \}\)/
\#\#\#\#\#

When I run this with 'use re "debugcolor";' commented out\, I get this&#8203;:

\#\#\#\#\#
$ perl 117917\_regex\.pl 
Use of uninitialized value $\+\{"delim"\} in die at \(re\_eval 2\) line 1\.
Died at \(re\_eval 2\) line 1\.
\#\#\#\#\#

When I activate 'use re "debugcolor";'\, the program ends with a bus
error \(see attachment\)\.

Now\, I have never actually used named captures in regular expressions\,
nor have I ever used 'use re' anything\.  So I can't claim to say
anything about the possible interaction of these features\.

But if this were my program I'd deal \*first\* with the warning and error
message thrown when you run without 'debugcolor'\.  I would then try to
build up the pattern match in small pieces\, with and without
'debugcolor'\.  The objective would be to determine what tips the program
into a segfault or bus error\.

I don't see any evidence yet that there is a problem with the Perl 5
core distribution\, which is the subject of this mailing list\.  You might
be better off taking this problem first to a place such as
perlmonks\.org\.  If discussion there points to the Perl 5 regex engine as
the problem \-\- as distinct from use of that engine \-\- then we could
resume discussion here\.

My two cents\.

Thank you very much\.
Jim Keenan

**Comment by p5pRT** [comment]
### From @jkeenan

$ perl 117917\_regex\.pl 
Compiling REx "\(?\<stringstart>\(?&#8203;:q\(?&#8203;:\[xwrq\]\)?\\s\*\)?\)\(?\<delim>\(??\{ \(exists \($"\.\.\.
Compiling REx "q"
Final program&#8203;:
&nbsp;   1&#8203;: EXACT \<q> \(3\)
&nbsp;   3&#8203;: END \(0\)
anchored "q" at 0 \(checking anchored isall\) minlen 1 
Final program&#8203;:
&nbsp;   1&#8203;: OPEN1 'stringstart' \(3\)
&nbsp;   3&#8203;:   CURLYX\[0\] \{0\,1\} \(23\)
&nbsp;   5&#8203;:     EXACT \<q> \(7\)
&nbsp;   7&#8203;:     CURLY \{0\,1\} \(20\)
&nbsp;   9&#8203;:       ANYOF\[qrwx\]\[\] \(0\)
&nbsp;  20&#8203;:     STAR \(22\)
&nbsp;  21&#8203;:       SPACE \(0\)
&nbsp;  22&#8203;:   WHILEM \(0\)
&nbsp;  23&#8203;:   NOTHING \(24\)
&nbsp;  24&#8203;: CLOSE1 'stringstart' \(26\)
&nbsp;  26&#8203;: OPEN2 'delim' \(28\)
&nbsp;  28&#8203;:   LOGICAL\[2\] \(29\)
&nbsp;  29&#8203;:   EVAL \(31\)
&nbsp;  31&#8203;: CLOSE2 'delim' \(33\)
&nbsp;  33&#8203;: MINMOD \(34\)
&nbsp;  34&#8203;: STAR \(36\)
&nbsp;  35&#8203;:   REG\_ANY \(0\)
&nbsp;  36&#8203;: UNLESSM\[\-1\] \(42\)
&nbsp;  38&#8203;:   EXACT \<\\\\> \(40\)
&nbsp;  40&#8203;:   SUCCEED \(0\)
&nbsp;  41&#8203;: TAIL \(42\)
&nbsp;  42&#8203;: LOGICAL\[2\] \(43\)
&nbsp;  43&#8203;: EVAL \(45\)
&nbsp;  45&#8203;: END \(0\)
minlen 0 with eval 
Matching REx "\(?\<stringstart>\(?&#8203;:q\(?&#8203;:\[xwrq\]\)?\\s\*\)?\)\(?\<delim>\(??\{ \(exists \($"\.\.\. against "qq%"hello%""
&nbsp;   0 \<qq"hello">|  1&#8203;:OPEN1 'stringstart'\(3\)
&nbsp;   0 \<qq"hello">|  3&#8203;:CURLYX\[0\] \{0\,1\}\(23\)
&nbsp;   0 \<qq"hello">| 22&#8203;:  WHILEM\(0\)
&nbsp;                                    whilem&#8203;: matched 0 out of 0\.\.1
&nbsp;   0 \<qq"hello">|  5&#8203;:    EXACT \<q>\(7\)
&nbsp;   1 \<qq"hello">|  7&#8203;:    CURLY \{0\,1\}\(20\)
&nbsp;                                      ANYOF\[qrwx\]\[\] can match 1 times out of 1\.\.\.
&nbsp;   2 \<qq"hello">| 20&#8203;:      STAR\(22\)
&nbsp;                                        SPACE can match 0 times out of 2147483647\.\.\.
&nbsp;   2 \<qq"hello">| 22&#8203;:        WHILEM\(0\)
&nbsp;                                          whilem&#8203;: matched 1 out of 0\.\.1
&nbsp;   2 \<qq"hello">| 23&#8203;:          NOTHING\(24\)
&nbsp;   2 \<qq"hello">| 24&#8203;:          CLOSE1 'stringstart'\(26\)
&nbsp;   2 \<qq"hello">| 26&#8203;:          OPEN2 'delim'\(28\)
&nbsp;   2 \<qq"hello">| 28&#8203;:          LOGICAL\[2\]\(29\)
&nbsp;   2 \<qq"hello">| 29&#8203;:          EVAL\(31\)
Guessing start of match in sv for REx "q" against "qq"
Found anchored substr "q" at offset 0\.\.\.
Guessed&#8203;: match at offset 0
Matching embedded REx "%"|'" against "%"hello%""
&nbsp;   2 \<qq"hello">|  1&#8203;:            TRIE\-EXACT\["'\]\(7\)
&nbsp;   2 \<qq"hello">|                State&#8203;:    1 Accepted&#8203;: N Charid&#8203;:  1 CP&#8203;:  22 After State&#8203;:    2
&nbsp;   3 \<qq"hello">|                State&#8203;:    2 Accepted&#8203;: Y Charid&#8203;:  0 CP&#8203;:   0 After State&#8203;:    0
&nbsp;                                              got 1 possible matches
&nbsp;                                              TRIE matched word \#1\, continuing
Bus error

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @tonycoz

On Wed\, May 08\, 2013 at 06&#8203;:14&#8203;:45PM \-0700\, James E Keenan via RT wrote&#8203;:
> The program as I have extracted it from your attachment&#8203;:
> 
> \#\#\#\#\#
> $ cat 117917\_regex\.pl 
> use strict;
> use warnings;
> 
> use re "debugcolor"; 
> 
> sub getDel \{
>           my $z = shift;
>           die $z;
>           return '\)' if $z eq '\(';
>           return '>' if $z eq '\<';
>           return '\]' if $z eq '\[';
>           return '\}' if $z eq '\{';
>           return $z;
> \} 
> 
> 'qq"hello"' =~ /\(?\<stringstart>\(?&#8203;:q\(?&#8203;:\[xwrq\]\)?\\s\*\)?\)\(?\<delim>\(??\{
> \(exists \($\+\{stringstart\}\) && $\+\{stringstart\} =~ m\#q\#\) ? q\#"|'\# : '\\\\w'
> \}\)\)\.\*?\(?\<\!\\\\\)\(??\{ die $\+\{delim\}; defined $1 ? quotemeta\(getDel\($1\)\) :
> '\(?\!\)' \}\)/
> \#\#\#\#\#

> I don't see any evidence yet that there is a problem with the Perl 5
> core distribution\, which is the subject of this mailing list\.  You might
> be better off taking this problem first to a place such as
> perlmonks\.org\.  If discussion there points to the Perl 5 regex engine as
> the problem \-\- as distinct from use of that engine \-\- then we could
> resume discussion here\.

Given it's perl code using only core modules\, the match is against a
short string\, and it crashes a with a Bus error\, I'd call this a bug
in perl\.

That said\, blead doesn't crash \(but still has an undefined $\+\{delim\}\)\.

Tony

**Comment by p5pRT** [comment]
### From @nwc10

On Thu\, May 09\, 2013 at 11&#8203;:32&#8203;:19AM \+1000\, Tony Cook wrote&#8203;:

> Given it's perl code using only core modules\, the match is against a
> short string\, and it crashes a with a Bus error\, I'd call this a bug
> in perl\.
> 
> That said\, blead doesn't crash \(but still has an undefined $\+\{delim\}\)\.

Tweaking the test case to this \(warn\, instead of die\)&#8203;:

$ cat \.\./117917\.pl
\#\!\./perl
use strict;
use warnings;

use re "debugcolor"; 

sub getDel \{
&nbsp;          my $z = shift;
&nbsp;          warn $z;
&nbsp;          return '\)' if $z eq '\(';
&nbsp;          return '>' if $z eq '\<';
&nbsp;          return '\]' if $z eq '\[';
&nbsp;          return '\}' if $z eq '\{';
&nbsp;          return $z;
\} 

'qq"hello"' =~ /\(?\<stringstart>\(?&#8203;:q\(?&#8203;:\[xwrq\]\)?\\s\*\)?\)\(?\<delim>\(??\{
\(exists \($\+\{stringstart\}\) && $\+\{stringstart\} =~ m\#q\#\) ? q\#"|'\# : '\\\\w'
\}\)\)\.\*?\(?\<\!\\\\\)\(??\{ warn $\+\{delim\}; defined $1 ? quotemeta\(getDel\($1\)\) :
'\(?\!\)' \}\)/


bisecting it with

$ bisect\.pl \-\-expect\-fail \-\-start v5\.16\.0 \.\./117917\.pl

says

8a45afe535d962511dc34619dcdb405aeff849da is the first bad commit
commit 8a45afe535d962511dc34619dcdb405aeff849da
Author&#8203;: David Mitchell \<davem@&#8203;iabyn\.com>
Date&#8203;:   Fri Oct 21 15&#8203;:00&#8203;:47 2011 \+0100

&nbsp;    unlink re\_eval code blocks from op list

&nbsp;    In the list of ops generated by something like /abc\(?\{\.\.\.\}\)def/\,

&nbsp;        const\(abc\)
&nbsp;        null/special
&nbsp;            \.\.\.
&nbsp;        const\(\.\.\.\)
&nbsp;        const\(def\)

&nbsp;    link the list\, but skip the DO blocks\. This means that for the runtime
&nbsp;    case\, we no longer need the temporary measure of deleting the DO blocks\,
&nbsp;    and it will facilitate the next step of handling literal code at runtime\,
&nbsp;    i\.e\. /$runtime\(?\{\.\.\.\}\)/\.

:100644 100644 1d7a7fde3ab4a110f7a6c1d540cddc6c02c3dc81 75667dff430ffb0ad56549aa1a0e4ffffac67e0f M      op\.c
bisect run success
That took 852 seconds


ie "bad" means that that commit fixed the bus error\.

Nicholas Clark

**Comment by p5pRT** [comment]
### From @iabyn

On Thu\, May 09\, 2013 at 11&#8203;:32&#8203;:19AM \+1000\, Tony Cook wrote&#8203;:
> That said\, blead doesn't crash \(but still has an undefined $\+\{delim\}\)\.

This can be reduced to;

'ab' =~ /^\(a\(?\{ 'x' =~ m\{x\}\}\)b\)\(?\{ warn "inner undef\!\\n" unless defined $1 \}\)/;
warn "outer undef\!\\n" unless defined $1;

which produces&#8203;:

&nbsp;    inner undef\!

removing the match from inside the \(?\{\}\) makes the issue go away\, so it looks
like the regex state isn't being saved/restored properly on re\-entrancy\.

I'm looking into why\.




\-\- 
A walk of a thousand miles begins with a single step\.\.\.
then continues for another 1\,999\,999 or so\.

**Comment by p5pRT** [comment]
### From @cpansprout

On Thu May 09 08&#8203;:54&#8203;:30 2013\, davem wrote&#8203;:
> On Thu\, May 09\, 2013 at 11&#8203;:32&#8203;:19AM \+1000\, Tony Cook wrote&#8203;:
> > That said\, blead doesn't crash \(but still has an undefined
> $\+\{delim\}\)\.
> 
> This can be reduced to;
> 
> 'ab' =~ /^\(a\(?\{ 'x' =~ m\{x\}\}\)b\)\(?\{ warn "inner undef\!\\n" unless
> defined $1 \}\)/;
> warn "outer undef\!\\n" unless defined $1;
> 
> which produces&#8203;:
> 
>     inner undef\!
> 
> removing the match from inside the \(?\{\}\) makes the issue go away\, so
> it looks
> like the regex state isn't being saved/restored properly on re\-
> entrancy\.
> 
> I'm looking into why\.

I have beaten you to it\. :\-\)

Fixed in f5df269\.

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
@cpansprout - Status changed from 'open' to 'resolved'
